From 64e0c0fe372416bc7bd412d88e953f44dadf8a6c Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sun, 27 Apr 2014 16:50:19 +0000
Subject: [PATCH] [CONFIGURATION-572] Fixed a possible memory leak.

When a CombinedConfiguration is cleared it is now ensured that change listener
registrations are removed for all child configurations.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1590417 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration/CombinedConfiguration.java  | 16 ++++++++++++++++
 .../TestCombinedConfiguration.java            | 19 +++++++++++++++++++
 2 files changed, 35 insertions(+)

diff --git a/src/main/java/org/apache/commons/configuration/CombinedConfiguration.java b/src/main/java/org/apache/commons/configuration/CombinedConfiguration.java
index 239466cd18..37513eb37f 100644
--- a/src/main/java/org/apache/commons/configuration/CombinedConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration/CombinedConfiguration.java
@@ -655,6 +655,7 @@ public void configurationChanged(ConfigurationEvent event)
     @Override
     protected void clearInternal()
     {
+        unregisterListenerAtChildren();
         initChildCollections();
         invalidateInternal();
     }
@@ -971,6 +972,21 @@ private void unregisterListenerAt(Configuration configuration)
         }
     }
 
+    /**
+     * Removes this combined configuration as listener from all child
+     * configurations. This method is called on a clear() operation.
+     */
+    private void unregisterListenerAtChildren()
+    {
+        if (configurations != null)
+        {
+            for (ConfigData child : configurations)
+            {
+                unregisterListenerAt(child.getConfiguration());
+            }
+        }
+    }
+
     /**
      * Returns the number of child configurations in this combined
      * configuration. The internal list of child configurations is accessed
diff --git a/src/test/java/org/apache/commons/configuration/TestCombinedConfiguration.java b/src/test/java/org/apache/commons/configuration/TestCombinedConfiguration.java
index f434c486ba..32424c87cd 100644
--- a/src/test/java/org/apache/commons/configuration/TestCombinedConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration/TestCombinedConfiguration.java
@@ -18,6 +18,7 @@
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNotSame;
 import static org.junit.Assert.assertNull;
@@ -449,6 +450,24 @@ public void testClear()
         listener.checkEvent(3, 2);
     }
 
+    /**
+     * Tests whether the combined configuration removes itself as change
+     * listener from the child configurations on a clear operation. This test is
+     * related to CONFIGURATION-572.
+     */
+    @Test
+    public void testClearRemoveChildListener()
+    {
+        AbstractConfiguration child = setUpTestConfiguration();
+        config.addConfiguration(child);
+
+        config.clear();
+        for (ConfigurationListener listener : child.getConfigurationListeners())
+        {
+            assertNotEquals("Still registered", config, listener);
+        }
+    }
+
     /**
      * Prepares a test of the getSource() method.
      */
